src/telemetry_poller.erl

```diff
@@ -0,0 +1,407 @@
+list_measurements(Poller) ->
+    gen_server:call(Poller, get_measurements).
+
+-spec init(map()) -> {ok, state()}.
+init(Args) ->
+    schedule_measurement(0),
+    {ok,
+     #{measurements => maps:get(measurements, Args),
+       period => maps:get(period, Args)}}.
+
+%% @doc
+%% Returns a child spec for the poller for running under a supervisor.
+child_spec(Opts) ->
+    Id =
+        case proplists:get_value(name, Opts, telemetry_poller_default) of
+            Name when is_atom(Name) -> Name;
+            {global, Name} -> Name;
+            {via, _, Name} -> Name
+        end,
+
+    #{
+        id => Id,
+        start => {telemetry_poller, start_link, [Opts]}
+    }.
+
+parse_args(Args) ->
+    Measurements = proplists:get_value(measurements, Args, []),
+    validate_measurements(Measurements),
+    Period = proplists:get_value(period, Args, 5000),
+    validate_period(Period),
+    VMMeasurements = parse_vm_measurements(proplists:get_value(vm_measurements, Args, [])),
+
+    #{measurements => Measurements ++ VMMeasurements, period => Period}.
+
+-spec parse_vm_measurements([] | default) -> [vm_measurement()].
+parse_vm_measurements([]) ->
+    [];
+parse_vm_measurements(default) ->
+    parse_vm_measurements(?DEFAULT_VM_MEASUREMENTS);
+parse_vm_measurements(Measurements) ->
+    __Measurements = sets:from_list(Measurements),
+    lists:map(fun parse_vm_measurement/1, sets:to_list(__Measurements)).
+
+-spec parse_vm_measurement(measurement()) -> {telemetry_poller_vm, measurement(), list()} | no_return().
+parse_vm_measurement(Measurement) ->
+    case lists:member(Measurement, ?DEFAULT_VM_MEASUREMENTS) of
+        true -> vm_measurement(Measurement);
+        false -> erlang:error(badarg, Measurement)
+    end.
+
+-spec vm_measurement(atom()) -> measurement().
+vm_measurement(Function) ->
+    {telemetry_poller_vm, Function, []}.
+
+-spec schedule_measurement(non_neg_integer()) -> ok.
+schedule_measurement(CollectInMillis) ->
+    erlang:send_after(CollectInMillis, self(), collect), ok.
+
+-spec validate_period(term()) -> ok | no_return().
+validate_period(Period) when is_integer(Period), Period > 0 ->
+    ok;
+validate_period(Term) ->
+    erlang:error(badarg, Term).
+
+-spec validate_measurements(term()) -> ok | no_return().
+validate_measurements(Measurements) when is_list(Measurements) ->
+    lists:map(fun validate_measurement/1, Measurements);
+validate_measurements(Term) ->
+    erlang:error(badarg, Term).
+
+-spec validate_measurement(term()) -> ok | no_return().
+validate_measurement({M, F, A}) when is_atom(M), is_atom(F), is_list(A) ->
+    ok;
+validate_measurement(Term) ->
+    erlang:error(badarg, Term).
+
+-spec make_measurements_and_filter_misbehaving([measurement()]) -> [measurement()].
+make_measurements_and_filter_misbehaving(Measurements) ->
+    Results = lists:map(fun make_measurement/1, Measurements),
+    lists:filter(fun(error) -> false; (_) -> true end, Results).
```

Comment:
The map+filter could be done in a single pass with a comprehension. :)

Reply:
Nice call!